File: utils/signal_processing_options/signalprocessingoptions.cpp
/*Copyright 2016 SuperDARN*/
#include <boost/property_tree/ini_parser.hpp>
#include <boost/range/algorithm_ext/erase.hpp>
#include <boost/lexical_cast.hpp>
#include <string>
#include <vector>
#include <regex>
#include <iostream>
#include "utils/options/options.hpp"
#include "utils/signal_processing_options/signalprocessingoptions.hpp"

std::vector<uint32_t> split(const std::string str, const std::string regex_str)
{
    std::regex re(regex_str);
    std::vector<std::string> str_list(
    std::sregex_token_iterator(str.begin(), str.end(), re, -1),
    std::sregex_token_iterator()
  );

  std::vector<uint32_t> int_list;
  for (auto& item: str_list) {
    int_list.push_back(boost::lexical_cast<uint32_t>(item));
  }

  return int_list;
}

SignalProcessingOptions::SignalProcessingOptions() {
  Options::parse_config_file();

  main_antenna_count_ = boost::lexical_cast<uint32_t>(
                              config_pt.get<std::string>("main_antenna_count"));
  interferometer_antenna_count_ = boost::lexical_cast<uint32_t>(
                              config_pt.get<std::string>("interferometer_antenna_count"));

  // Parse N200 list and determine which antennas are active
  std::vector<uint32_t> main_antenna_vec;
  std::vector<uint32_t> intf_antenna_vec;
  auto n200_list = config_pt.get_child("n200s");
  for (auto n200 = n200_list.begin(); n200 != n200_list.end(); n200++) {
    std::string addr = "";
    bool rx = false;
    bool tx = false;
    bool rx_int = false;
    std::string main_antenna = "";
    std::string interferometer_antenna = "";
    // Iterate through all N200 parameters and store them in variables
    for (auto iter = n200->second.begin(); iter != n200->second.end(); iter++)
    {
      auto param = iter->first;
      if (param.compare("addr") == 0) {
        addr = iter->second.data();
      }
      else if (param.compare("rx") == 0) {
        rx = (iter->second.data().compare("true") == 0);
      }
      else if (param.compare("tx") == 0) {
        tx = (iter->second.data().compare("true") == 0);
      }
      else if (param.compare("rx_int") == 0) {
        rx_int = (iter->second.data().compare("true") == 0);
      }
      else if (param.compare("main_antenna") == 0) {
        main_antenna = iter->second.data();
      }
      else if (param.compare("interferometer_antenna") == 0) {
        interferometer_antenna = iter->second.data();
      }
      else {
        throw std::invalid_argument("Invalid N200 parameter in config file");
      }
    }
    if (rx || tx) {
      auto main_antenna_num = boost::lexical_cast<uint32_t>(main_antenna);
      main_antenna_vec.push_back(main_antenna_num);
    }
    if (rx_int) {
      auto int_antenna_num = boost::lexical_cast<uint32_t>(interferometer_antenna);
      intf_antenna_vec.push_back(int_antenna_num);
    }
  }
  std::sort(main_antenna_vec.begin(), main_antenna_vec.end());
  std::sort(intf_antenna_vec.begin(), intf_antenna_vec.end());
  std::string main_antenna_list = "";
  std::string interferometer_antenna_list = "";
  for (auto antenna_num : main_antenna_vec) {  // TODO: Create the attribute directly from the config file without the secondary function
    main_antenna_list = main_antenna_list + std::to_string(antenna_num) + ",";
  }
  for (auto intf_num : intf_antenna_vec) {
    interferometer_antenna_list = interferometer_antenna_list + std::to_string(intf_num) + ",";
  }
  main_antenna_list.pop_back();
  interferometer_antenna_list.pop_back();

  main_antennas = split(main_antenna_list, ",");
  interferometer_antennas = split(interferometer_antenna_list, ",");

  router_address = config_pt.get<std::string>("router_address");
  dsp_to_radctrl_identity = config_pt.get<std::string>("dsp_to_radctrl_identity");
  dsp_driver_identity = config_pt.get<std::string>("dsp_to_driver_identity");
  dsp_exphan_identity = config_pt.get<std::string>("dsp_to_exphan_identity");
  dsp_dw_identity = config_pt.get<std::string>("dsp_to_dw_identity");
  dspbegin_brian_identity = config_pt.get<std::string>("dspbegin_to_brian_identity");
  dspend_brian_identity = config_pt.get<std::string>("dspend_to_brian_identity");
  radctrl_dsp_identity = config_pt.get<std::string>("radctrl_to_dsp_identity");
  driver_dsp_identity = config_pt.get<std::string>("driver_to_dsp_identity");
  brian_dspbegin_identity = config_pt.get<std::string>("brian_to_dspbegin_identity");
  brian_dspend_identity = config_pt.get<std::string>("brian_to_dspend_identity");
  exphan_dsp_identity = config_pt.get<std::string>("exphan_to_dsp_identity");
  dw_dsp_identity = config_pt.get<std::string>("dw_to_dsp_identity");
  ringbuffer_name = config_pt.get<std::string>("ringbuffer_name");
}

std::vector<uint32_t> SignalProcessingOptions::get_main_antennas() const
{
  return main_antennas;
}

uint32_t SignalProcessingOptions::get_main_antenna_count() const
{
  return main_antenna_count;
}

std::vector<uint32_t> SignalProcessingOptions::get_interferometer_antennas() const
{
  return interferometer_antennas;
}

uint32_t SignalProcessingOptions::get_interferometer_antenna_count() const
{
  return interferometer_antenna_count;
}

std::string SignalProcessingOptions::get_router_address() const
{
  return router_address;
}

std::string SignalProcessingOptions::get_dsp_radctrl_identity() const
{
  return dsp_to_radctrl_identity;
}

std::string SignalProcessingOptions::get_dsp_driver_identity() const
{
  return dsp_driver_identity;
}

std::string SignalProcessingOptions::get_dsp_exphan_identity() const
{
  return dsp_exphan_identity;
}

std::string SignalProcessingOptions::get_dsp_dw_identity() const
{
  return dsp_dw_identity;
}

std::string SignalProcessingOptions::get_dspbegin_brian_identity() const
{
  return   dspbegin_brian_identity;
}

std::string SignalProcessingOptions::get_dspend_brian_identity() const
{
  return   dspend_brian_identity;
}

std::string SignalProcessingOptions::get_radctrl_dsp_identity() const
{
  return radctrl_dsp_identity;
}

std::string SignalProcessingOptions::get_driver_dsp_identity() const
{
  return driver_dsp_identity;
}

std::string SignalProcessingOptions::get_brian_dspbegin_identity() const
{
  return brian_dspbegin_identity;
}

std::string SignalProcessingOptions::get_brian_dspend_identity() const
{
  return brian_dspend_identity;
}

std::string SignalProcessingOptions::get_exphan_dsp_identity() const
{
  return exphan_dsp_identity;
}

std::string SignalProcessingOptions::get_dw_dsp_identity() const
{
  return dw_dsp_identity;
}

std::string SignalProcessingOptions::get_ringbuffer_name() const
{
  return ringbuffer_name;
}

File: utils/signal_processing_options/signal_processing_options.py
#!/usr/bin/python

# Copyright 2020 SuperDARN Canada
#
# signal_processing_options.py
# 2020-09-28
# options class for signal processing module

import json
import os

class SignalProcessingOptions(object):
    """
    Parses the options from the config file that are relevant to signal processing.

    """
    def __init__(self):
        super(SignalProcessingOptions, self).__init__()

        if not os.environ["BOREALISPATH"]:
            raise ValueError("BOREALISPATH env variable not set")
        config_path = os.environ["BOREALISPATH"] + "/config.ini"
        try:
            with open(config_path, 'r') as config_data:
                raw_config = json.load(config_data)
        except IOError:
            errmsg = 'Cannot open config file at {0}'.format(config_path)
            raise IOError(errmsg)

        self._router_address = raw_config["router_address"]
        self._dsp_radctrl_identity = raw_config["dsp_to_radctrl_identity"]
        self._dsp_driver_identity = raw_config["dsp_to_driver_identity"]
        self._dsp_exphan_identity = raw_config["dsp_to_exphan_identity"]
        self._dsp_dw_identity = raw_config["dsp_to_dw_identity"]
        self._dspbegin_brian_identity = raw_config["dspbegin_to_brian_identity"]
        self._dspend_brian_identity = raw_config["dspend_to_brian_identity"]
        self._radctrl_dsp_identity = raw_config["radctrl_to_dsp_identity"]
        self._driver_dsp_identity = raw_config["driver_to_dsp_identity"]
        self._brian_dspbegin_identity = raw_config["brian_to_dspbegin_identity"]
        self._brian_dspend_identity = raw_config["brian_to_dspend_identity"]
        self._exphan_dsp_identity = raw_config["exphan_to_dsp_identity"]
        self._dw_dsp_identity = raw_config["dw_to_dsp_identity"]
        self._ringbuffer_name = raw_config["ringbuffer_name"]
        self._main_antenna_count = int(raw_config["main_antenna_count"])
        self._intf_antenna_count = int(raw_config["interferometer_antenna_count"])
        self._main_antennas = []
        self._intf_antennas = []
        for n200 in raw_config["n200s"]:
            rx = bool(n200["rx"])
            tx = bool(n200["tx"])
            rx_int = bool(n200["rx_int"])
            if rx or tx:
                main_antenna_num = int(n200["main_antenna"])
                self._main_antennas.append(main_antenna_num)
            if rx_int:
                intf_antenna_num = int(n200["interferometer_antenna"])
                self._intf_antennas.append(intf_antenna_num)
        self._main_antennas.sort()
        self._intf_antennas.sort()


    @property
    def router_address(self):
        """
        Gets the socket address of the router that routes interprocess messages.

        :return:    socket address of the router that routes interprocess messages.
        :rtype:     str
        """
        return self._router_address

    @property
    def dsp_radctrl_identity(self):
        """
        Gets the socket identity for dsp to radar control.

        :returns:   dsp to radar control identity.
        :rtype:     str
        """

        return self._dsp_radctrl_identity

    @property
    def dsp_driver_identity(self):
        """
        Gets the socket identity for dsp to driver.

        :returns:   dsp to driver identity.
        :rtype:     str
        """
        return self._dsp_driver_identity

    @property
    def dsp_exphan_identity(self):
        """
        Gets the socket identity for dsp to exphan.

        :returns:   dsp to exphan identity.
        :rtype:     str
        """
        return self._dsp_exphan_identity

    @property
    def dsp_dw_identity(self):
        """
        Gets the socket identity for dsp to dw.

        :returns:   dsp to dw identity.
        :rtype:     str
        """
        return self._dsp_dw_identity

    @property
    def dspbegin_brian_identity(self):
        """
        Gets the socket identity for dspbegin to brian.

        :returns:   dspbegin to brian identity.
        :rtype:     str
        """
        return self._dspbegin_brian_identity

    @property
    def dspend_brian_identity(self):
        """
        Gets the socket identity for dspend to brian.

        :returns:   dspend to brian identity.
        :rtype:     str
        """
        return self._dspend_brian_identity

    @property
    def radctrl_dsp_identity(self):
        """
        Gets the socket identity for radctrl to dsp.

        :returns:   radctrl to dsp identity.
        :rtype:     str
        """
        return self._radctrl_dsp_identity

    @property
    def driver_dsp_identity(self):
        """
        Gets the socket identity for driver to dsp.

        :returns:   driver to dsp identity.
        :rtype:     str
        """
        return self._driver_dsp_identity

    @property
    def brian_dspbegin_identity(self):
        """
        Gets the socket identity for brian to dspbegin.

        :returns:   brian to dspbegin identity.
        :rtype:     str
        """
        return self._brian_dspbegin_identity

    @property
    def brian_dspend_identity(self):
        """
        Gets the socket identity for brian to dspend.

        :returns:   Brian to dspend identity.
        :rtype:     str
        """
        return self._brian_dspend_identity

    @property
    def exphan_dsp_identity(self):
        """
        Gets the socket identity for exphan to dsp.

        :returns:   Exphan to dsp identity.
        :rtype:     str
        """
        return self._exphan_dsp_identity

    @property
    def dw_dsp_identity(self):
        """
        Gets the socket identity for dw to dsp.

        :returns:   dw to dsp identity.
        :rtype:     str
        """
        return self._dw_dsp_identity

    @property
    def ringbuffer_name(self):
        """
        Gets the shared memory ringbuffer name.

        :returns:   The ringbuffer name.
        :rtype:     str
        """
        return self._ringbuffer_name

    @property
    def main_antenna_count(self):
        """
        Gets the main antenna count.

        :returns:   Number of main antennas.
        :rtype:     int
        """
        return self._main_antenna_count

    @property
    def intf_antenna_count(self):
        """
        Gets the intf antenna count.

        :returns:   Number of intf antennas.
        :rtype:     int
        """
        return self._intf_antenna_count


if __name__ == '__main__':
    SignalProcessingOptions()
